go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb

```diff
@@ -235,12 +235,26 @@ def replace_directive_substitutions(manifest)
               (manifest["Replace"] || []).
                 map { |r| r["New"]["Path"] }.
                 compact.
-                select { |p| p.start_with?(".") || p.start_with?("/") }.
+                select { |p| external_path?(p) }.
                 map { |p| [p, "./" + Digest::SHA2.hexdigest(p)] }.
                 to_h
             end
         end
 
+        # returns true if the provided path does not exist in the current checkout
+        def external_path?(path)
+          return true if path.start_with?("/")
+          return false unless path.start_with?(".")
```

Comment:
I'm trying to understand these lines. How does these lines allow to understand whether the path is not inside the repository? I think a `/foo/bar/repo_path/qux` still could point into a folder that is inside the repository or `./foo` could point to a folder inside a repository. I wonder if `external_path` is meaning something else.

Reply:
> I think a /foo/bar/repo_path/qux still could point into a folder that is inside the repository

This is unlikely. Since Dependabot clones the repository to a temporary directory, an absolute path inside the repository would have to be `/tmp/dependabot-${random}`, where `${random}` varies between runs. It _could_ happen, but is so unlikely we can save an `fstat()` by not even checking.

> ./foo could point to a folder inside a repository.

This should be catching that case. Per https://golang.org/ref/mod#go-mod-file-replace , we know only `./foo` and `../foo` are valid so the `path.start_with?(".")` check attempts to check that.
If a replacement path _isn't_ a relative path, then Dependabot shouldn't replace it - the `go.mod` file is invalid so the most useful thing to do is return the error from the `go` SDK:
```
go: errors parsing go.mod:
/Users/thepwagner/tmp/go_mod_test/go.mod:7: replacement module without version must be directory path (rooted or starting with ./ or ../)
```


> I wonder if external_path is meaning something else.

Good shout - I've renamed to more closely what this is doing: deciding if this path should be stubbed / `#stub_replace_path?`